pkg/block/blocktest/multipart.go

```diff
@@ -0,0 +1,315 @@
+)
+
+// Parameterized test of the Multipart Upload APIs. After successful upload we Get the result and compare to the original
+func testAdapterMultipartUpload(t *testing.T, adapter block.Adapter, storageNamespace string) {
+	ctx := context.Background()
+	parts, full := createMultipartContents()
+
+	cases := []struct {
+		name string
+		path string
+	}{
+		{"simple", "abc"},
+		{"nested", "foo/bar"},
+	}
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			blockstoreType := adapter.BlockstoreType()
+			obj := block.ObjectPointer{
+				StorageNamespace: storageNamespace,
+				Identifier:       c.path,
+				IdentifierType:   block.IdentifierTypeRelative,
+			}
+
+			verifyListInvalid(t, ctx, adapter, obj, "invalidId")
+
+			resp, err := adapter.CreateMultiPartUpload(ctx, obj, nil, block.CreateMultiPartUploadOpts{})
+			require.NoError(t, err)
+
+			multiParts := uploadParts(t, ctx, adapter, obj, resp.UploadID, parts)
+
+			// List parts after upload
+			listResp, err := adapter.ListParts(ctx, obj, resp.UploadID, block.ListPartsOpts{})
+			if blockstoreType != block.BlockstoreTypeS3 {
+				require.ErrorIs(t, err, block.ErrOperationNotSupported)
+			} else {
+				require.NoError(t, err)
+				require.Equal(t, len(parts), len(listResp.Parts))
+				for i, part := range listResp.Parts {
+					require.Equal(t, multiParts[i].PartNumber, part.PartNumber)
+					require.Equal(t, int64(len(parts[i])), part.Size)
+					require.Equal(t, multiParts[i].ETag, part.ETag)
+					require.False(t, listResp.IsTruncated)
+				}
+			}
+
+			// List parts partial
+			const maxPartsConst = 2
+			maxParts := int32(maxPartsConst)
+			listResp, err = adapter.ListParts(ctx, obj, resp.UploadID, block.ListPartsOpts{MaxParts: &maxParts})
+			if blockstoreType != block.BlockstoreTypeS3 {
+				require.ErrorIs(t, err, block.ErrOperationNotSupported)
+			} else {
+				require.NoError(t, err)
+				require.Equal(t, int(maxParts), len(listResp.Parts))
+				require.True(t, listResp.IsTruncated)
+				require.Equal(t, strconv.Itoa(int(maxParts)), *listResp.NextPartNumberMarker)
+			}
+
+			_, err = adapter.CompleteMultiPartUpload(ctx, obj, resp.UploadID, &block.MultipartUploadCompletion{
+				Part: multiParts,
+			})
+			require.NoError(t, err)
+
+			verifyListInvalid(t, ctx, adapter, obj, resp.UploadID)
+
+			getAndCheckContents(t, ctx, adapter, full, obj)
+		})
+	}
+}
+
+// Test aborting a multipart upload
+func testAdapterAbortMultipartUpload(t *testing.T, adapter block.Adapter, storageNamespace string) {
+	ctx := context.Background()
+	parts, _ := createMultipartContents()
+	obj, _ := objPointers(storageNamespace)
+	uploadID := "foobar"
+
+	err := adapter.AbortMultiPartUpload(ctx, obj, uploadID)
+	if err != nil {
+		if adapter.BlockstoreType() == block.BlockstoreTypeLocal {
+			require.ErrorContains(t, err, "invalid upload id")
+		} else {
+			require.ErrorContains(t, err, "404")
+		}
```

Comment:
It's not recommended to check the returned error.
Basically, if the adapter was expected to return a specific error I would understand, but in this case we are testing the text returned from the underlying storage